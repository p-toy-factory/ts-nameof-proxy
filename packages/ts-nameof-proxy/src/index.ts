export { nameOf } from "./name-of";
export { namesOf as unstable_namesOf } from "./names-of";
export { pathOf } from "./path-of";
export { pathsOf as unstable_pathsOf } from "./paths-of";
export { separatedPathOf } from "./separated-path-of";
export { separatedPathsOf as unstable_separatedPathsOf } from "./separated-paths-of";
export type { CallBackForPropertyAccess, NameSelector } from "./types";
