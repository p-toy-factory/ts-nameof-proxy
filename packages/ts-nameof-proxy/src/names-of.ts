import { pathsOf } from "./paths-of";
import { NameSelector } from "./types";
import { last } from "./utils";

/**
 * @example
 * namesOf(student, (s) => (s.age, s.name.length));  // ["age", "length"]
 * namesOf<Student>((s) => (s.age, s.name.length));  // ["age", "length"]
 */
export function namesOf<T>(callback: NameSelector<T>): string[];
export function namesOf<T>(obj: T, callback: NameSelector<T>): string[];

export function namesOf<T>(
	arg1: T | NameSelector<T>,
	arg2?: NameSelector<T>
): string[] {
	return pathsOf(arg1, arg2).map(last);
}
