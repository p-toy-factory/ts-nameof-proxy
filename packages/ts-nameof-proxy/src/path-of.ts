import { pathsOf } from "./paths-of";
import { NameSelector } from "./types";

/**
 * @example
 * pathOf(student, (s) => s.age);          // ["age"]
 * pathOf(student, (s) => s.name.length);  // ["name", "length"]
 * pathOf<Student>((s) => s.name.length);  // ["name", "length"]
 */
export function pathOf<T>(selector: NameSelector<T>): string[];
export function pathOf<T>(obj: T, selector?: NameSelector<T>): string[];

export function pathOf<T>(
	objOrSelector: T | NameSelector<T>,
	selectorOrNil?: NameSelector<T>
): string[] {
	return pathsOf(objOrSelector, selectorOrNil)[0] ?? [];
}
