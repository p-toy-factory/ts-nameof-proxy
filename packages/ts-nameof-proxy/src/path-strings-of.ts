import { pathsOf } from "./paths-of";
import { NameSelector } from "./types";

/**
 * @example
 * pathStringsOf(student, (s) => (s.name, s.name.firstName[0])); // ["['name']", "['name']['firstName']['0']"]
 * pathStringsOf<Student>((s) => (s.name, s.name.firstName[0])); // ["['name']", "['name']['firstName']['0']"]
 */
export function pathStringsOf<T>(callback: NameSelector<T>): string[];
export function pathStringsOf<T>(obj: T, callback: NameSelector<T>): string[];

export function pathStringsOf<T>(
	arg1: T | NameSelector<T>,
	arg2?: NameSelector<T>
): string[] {
	return pathsOf(arg1, arg2).map(mapper);
}

function mapper(separatedPath: string[]) {
	return `['${separatedPath.join("']['")}']`;
}
