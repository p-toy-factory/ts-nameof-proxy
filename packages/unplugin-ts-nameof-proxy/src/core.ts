import generate from "@babel/generator";
import * as parser from "@babel/parser";
import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { NodePath } from "babel__traverse";
import { equals, head, identity, last, length, map, pipe } from "rambda";
import { match, P } from "ts-pattern";

/**
 * @example
 * ```
 * const ast = parse("a.b.c");
 * getInitialObjectNameOfMemberExpression(ast) === "a";
 * ```
 */
function getInitialObjectNameOfMemberExpression(
	node: t.MemberExpression,
): string {
	return match(node.object)
		.when(t.isIdentifier, (object) => object.name)
		.when(t.isMemberExpression, getInitialObjectNameOfMemberExpression)
		.otherwise(() => {
			throw new TypeError(`Unexpected node type: ${node.object.type}`);
		});
}

function propertiesToAST(
	properties: string | string[] | string[][],
): t.StringLiteral | t.ArrayExpression {
	return match(properties)
		.with(P.string, t.stringLiteral)
		.otherwise(pipe(map(propertiesToAST), t.arrayExpression));
}

/**
 * @todo Extract to a separate package
 * @example
 * ```
 * toPathString(["name", "length", "toString"]) === "['name']['length']['toString']"
 * ```
 */
function toPathString(path: string[]) {
	return `['${path.join("']['")}']`;
}

const pathsOf = identity<string[][]>;
const pathOf = pipe(pathsOf, head<string[][]>);
const pathStringsOf = pipe(pathsOf, map(toPathString));
const pathStringOf = pipe(pathStringsOf, head<string[]>);
const namesOf = pipe(pathsOf, map(last<string[]>));
const nameOf = pipe(namesOf, head<string[]>);

const propertiesMappers = {
	nameOf,
	namesOf,
	pathOf,
	pathsOf,
	pathStringOf,
	pathStringsOf,
} satisfies Record<
	string,
	(paths: string[][]) => string | string[] | string[][]
>;

const exportedFunctionNames = Object.keys(propertiesMappers) as Array<
	keyof typeof propertiesMappers
>;

function isCalleeIdentifierToTransform(
	path: NodePath<t.Identifier>,
	parentPath: NodePath<t.Node>,
	moduleSource: string,
): parentPath is NodePath<t.CallExpression> {
	if (path.parentPath !== parentPath) {
		throw new TypeError("path.parentPath !== parentPath");
	}
	return (
		path.parentPath.isCallExpression() &&
		exportedFunctionNames.some((name) =>
			path.referencesImport(moduleSource, name),
		)
	);
}

const isEmptyArray = pipe(length, equals(0));

export function tryGetCallExpressionToTransform(path: NodePath<t.Identifier>) {
	const { parentPath } = path;
	if (
		parentPath.isCallExpression() &&
		exportedFunctionNames.some((name) =>
			path.referencesImport("ts-nameof-proxy", name),
		) &&
		// Avoid the expressions like `String(pathStringOf)`
		parentPath.node.callee === path.node
	) {
		return {
			callExpressionPath: parentPath,
			functionName: path.node.name as keyof typeof propertiesMappers,
		};
	}
	return undefined;
}

export function tryGetInlineArrowFunctionSelector(
	callExpressionPath: NodePath<t.CallExpression>,
) {
	const args = callExpressionPath.node.arguments;
	const selector = args[1] ?? args[0];
	if (
		selector &&
		(t.isFunctionExpression(selector) || t.isArrowFunctionExpression(selector))
	) {
		return {
			selector,
		};
	}
	return undefined;
}

export function getMemberExpressionsIfCanBeTransformed(
	selector: t.FunctionExpression | t.ArrowFunctionExpression,
) {
	if (!t.isArrowFunctionExpression(selector)) {
		return undefined;
	}

	const selectorParam = selector.params[0];
	if (!t.isIdentifier(selectorParam)) {
		return undefined;
	}

	const memberExpressions = (() => {
		switch (selector.body.type) {
			case "MemberExpression":
				return [selector.body];

			case "SequenceExpression": {
				if (
					selector.body.expressions.length > 0 &&
					selector.body.expressions.every((expr) => t.isMemberExpression(expr))
				) {
					return selector.body.expressions as t.MemberExpression[];
				}
			}

			default:
				// The body of node of identify function is a identifier node
				return [];
		}
	})();

	// ✅ nameOf(p => p.name.length)
	// ❌ nameOf(p => a.name.length)
	const proxyName = selectorParam.name;
	const hasUnexpectedSelector = memberExpressions.some((expr) => {
		try {
			const initialObjectName = getInitialObjectNameOfMemberExpression(expr);
			return proxyName !== initialObjectName;
		} catch (error) {
			console.log(error);
			return true;
		}
	});

	if (hasUnexpectedSelector || memberExpressions.length === 0) {
		return undefined;
	}

	return {
		memberExpressions,
	};
}

function getMemberExpressionPropertyString(expr: t.MemberExpression): string {
	const { property } = expr;
	switch (property.type) {
		case "Identifier":
			return property.name;

		case "NumericLiteral":
		case "StringLiteral":
			return property.value.toString();

		default:
			throw new TypeError(`Unexpected property type: '${property.type}'`);
	}
}

function getMemberExpressionPath(expr: t.MemberExpression): string[] {
	const name = getMemberExpressionPropertyString(expr);

	switch (expr.object.type) {
		case "MemberExpression":
			return [...getMemberExpressionPath(expr.object), name];

		default:
			return [name];
	}
}

function identifierVisitor(path: NodePath<t.Identifier>) {
	const callExpressionPathResult = tryGetCallExpressionToTransform(path);
	if (!callExpressionPathResult) {
		return;
	}
	const { callExpressionPath, functionName } = callExpressionPathResult;

	const selectorResult = tryGetInlineArrowFunctionSelector(callExpressionPath);
	if (!selectorResult) {
		return;
	}
	const { selector } = selectorResult;

	const memberExpressionsResult =
		getMemberExpressionsIfCanBeTransformed(selector);
	if (!memberExpressionsResult) {
		return;
	}
	const { memberExpressions } = memberExpressionsResult;

	let hasUnexpectedNodeType = false;

	const propertiesArrays = memberExpressions.reduce<string[][]>(
		(propertiesArrays, expr) => {
			try {
				const properties = getMemberExpressionPath(expr);
				return [...propertiesArrays, properties];
			} catch (error) {
				console.log(error);
				hasUnexpectedNodeType = true;
				return propertiesArrays;
			}
		},
		[],
	);

	if (hasUnexpectedNodeType) {
		return;
	}

	const mapper = propertiesMappers[functionName];
	pipe(mapper, propertiesToAST, (ast) => callExpressionPath.replaceWith(ast))(
		propertiesArrays,
	);
}

export function transform(code: string): string {
	const ast = parser.parse(code, {
		sourceType: "module",
	});

	traverse(ast, {
		Identifier(path) {
			identifierVisitor(path);
		},
	});

	return generate(ast).code;
}
