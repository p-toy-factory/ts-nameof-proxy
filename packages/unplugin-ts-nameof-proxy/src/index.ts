import { createUnplugin } from "unplugin";

const plugins = createUnplugin((options) => {
	return {
		name: "unplugin-ts-nameof-proxy",
		transformInclude(id) {
			return /\.[jt]sx?$/.test(id);
		},
		transform(code) {
			return "";
		},
	};
});
