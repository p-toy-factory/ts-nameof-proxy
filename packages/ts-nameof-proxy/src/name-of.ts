import { namesOf } from "./names-of";
import { NameSelector } from "./types";

/**
 * @example
 * nameOf(student, (s) => s.age);         // "age"
 * nameOf(student, (s) => s.name.length); // "length"
 * nameOf<Student>((s) => s.name.length); // "length"
 */
export function nameOf<T>(callback: NameSelector<T>): string;
export function nameOf<T>(obj: T, callback: NameSelector<T>): string;

export function nameOf<T>(
	arg1: T | NameSelector<T>,
	arg2?: NameSelector<T>
): string {
	// @ts-ignore
	const names = namesOf(arg1, arg2);
	if (names.length === 0) {
		throw new Error("ts-nameof-proxy: No properties were read.");
	}
	return names[0];
}
