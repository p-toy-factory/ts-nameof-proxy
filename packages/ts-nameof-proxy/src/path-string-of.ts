import { pathStringsOf } from "./path-strings-of";
import { NameSelector } from "./types";

/**
 * @example
 * pathStringOf(student, (s) => s.name.firstName[0]); // "['name']['firstName']['0']"
 * pathStringOf<Student>((s) => s.name.firstName[0]); // "['name']['firstName']['0']"
 */
export function pathStringOf<T>(selector: NameSelector<T>): string;
export function pathStringOf<T>(obj: T, selector: NameSelector<T>): string;

export function pathStringOf<T>(
	arg1: T | NameSelector<T>,
	arg2?: NameSelector<T>
): string {
	// @ts-ignore
	const separatedPath = pathStringsOf(arg1, arg2);
	if (separatedPath.length === 0) {
		throw new Error("ts-nameof-proxy: No properties were read.");
	}
	return separatedPath[0];
}
