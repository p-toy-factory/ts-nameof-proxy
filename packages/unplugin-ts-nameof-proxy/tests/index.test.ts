import { expect, test } from "vitest";
import { transform } from "../src/core";
import { codeToOptimize, codeToPreserve } from "./fixtures";

test("Expect code will be optimized", () => {
	expect(transform(codeToOptimize)).toMatchSnapshot();
});

test("Expect code won't be optimized", () => {
	expect(transform(codeToPreserve)).toMatchSnapshot();
});
