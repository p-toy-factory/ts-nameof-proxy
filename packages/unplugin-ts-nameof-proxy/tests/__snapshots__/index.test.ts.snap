// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expect code will be optimized 1`] = `
"import { nameOf, namesOf, pathOf, pathsOf, pathStringOf, pathStringsOf } from "ts-nameof-proxy";
"length";
["toString", "length"];
["name", "length", "toString"];
[["name", "length", "toString"]];
[["name", "length", "toString"], ["name", "length"]];
"['name']['length']['toString']";
["['name']['length']['toString']", "['name']['length']"];"
`;

exports[`Expect code won't be optimized 1`] = `
"import { nameOf, namesOf, pathOf, pathsOf, pathStringOf, pathStringsOf } from "ts-nameof-proxy";
nameOf(f => f);
nameOf(f => {
  return f;
});
String(pathsOf);
pathsOf(["2"][1][""][index].length);
pathsOf(f => [f.name.length.toString, f.name.length]);
pathsOf(f => [f.name, f]);
pathsOf(f => [f.name, globalThis.fetch]);"
`;
